opus/application/static_media/js/browse.js

```diff
@@ -548,34 +521,94 @@ var o_browse = {
                     Left: 37
                     Space: 32 */
 
+                let tab = opus.getViewTab();
+                let viewNamespace = opus.getViewNamespace();
                 let detailOpusId;
+                let offset = 0;
+                let obsNum = $("#galleryViewContents .op-obs-direction a").data("obs");
+                let lastCachedObsNum = $(`${tab} .op-thumbnail-container`).last().data("obs");
                 // the || is for cross-browser support; firefox does not support keyCode
                 switch (e.which || e.keyCode) {
                     case 32:  // spacebar
-                        if (opus.metadataDetailOpusId !== "") {
+                        if (opus.metadataDetailOpusId !== "" && opus.metadataDetailOpusId !== undefined && !$("#galleryViewContents").hasClass("op-disabled")) {
                             o_browse.undoRangeSelect();
                             o_cart.toggleInCart(opus.metadataDetailOpusId);
                         }
                         break;
                     case 39:  // next
-                        detailOpusId = $("#galleryView").find(".op-next").data("id");
-                        o_browse.removeEditMetadataDetails();
-                        o_browse.loadPageIfNeeded("next", detailOpusId);
+                        obsNum++;
+                        // if this new observation is in the last row of cached observations, don't allow the user to arrow there
+                        if (lastCachedObsNum === viewNamespace.totalObsCount || obsNum + viewNamespace.galleryBoundingRect.x <= lastCachedObsNum) {
+                            detailOpusId = $("#galleryView").find(".op-next").data("id");
+                            o_browse.removeEditMetadataDetails();
+                            o_browse.loadPageIfNeeded("next", detailOpusId);
+                        } else {
+                            return false;
+                        }
                         break;
                     case 37:  // prev
+                        obsNum--;
```

Comment:
Why don't we check for obsNum < 1 here and return false like everywhere else in this switch?

Reply:
Even though the check is unnecessary as there are other checks in place for this, i have added a check to the code for consistency.